### PR TITLE
[refactoring] connecting simulation and execution APIs

### PR DESCRIPTION
**Pull request by georgemitenkov**
### Description

TLDR:; cherry pick from #10757 to refactor simulation flow.

This PR refactors Aptos VM simulation APIs. Previously, simulation has
multiple problems:

1. `AptosVM` was wrapped into `SimulationVM` with transaction validation
    and transaction execution APIs duplicated.
2. Code duplication resulted in divergence in behaviour, e.g., resource group
    cache was not flushed in simulation, etc.

This PR refactors simulation and introduces a wrapper: `AptosSimulationVM`,
which ensures encapsulation of `is_simulation` field, and exposes APIs to
simulate transactions.. Simulation call is resolved by calling
`execute_user_transaction`, so the same behaviour (if we do not consider
aggregators) as in standard execution is preserved.

### Test Plan

1. Added test to catch simulation of a txn with a valid signature.
2. Adapted simulation tests to fail on error.

### CODE REVIEW

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 1]
```diff
@@ -1125,6 +1122,47 @@ impl AptosVM {
         )
     }
 
+    // Called when the execution of the user transaction fails, in order to discard the
+    // transaction, or clean up the failed state.
+    fn on_user_transaction_execution_failure(
```
Comment (runtian-zhou) [inline aptos-move/aptos-vm/src/aptos_vm.rs:1127]: Should the execute script use the same cleanup function? Seems that it's only used for multi sig txn?

**Comment by vgao1996** [comment]
Quick question: which PR are you planning to land first, this one or https://github.com/aptos-labs/aptos-core/pull/10852?

**Comment by georgemitenkov** [comment]
@vgao1996 I was planning to land this one first, and then #10852. But these 2 can be reviewed independently I think

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/aptos_vm.rs:1127]: Which script execution do you refer to @runtian-zhou ? So far I use it in the end of `execute_user_transaction_impl`

File: api/src/transactions.rs [hunk 2]
```diff
@@ -1191,7 +1190,20 @@ impl TransactionsApi {
...
+
+        let output = vm
+            .simulate_signed_transaction(&txn, &state_view)
+            .map_err(|vm_status| {
+                SubmitTransactionError::internal_with_code(
+                    format!("Failed to simulate a transaction: {}", vm_status),
+                    AptosErrorCode::InternalError,
+                    &ledger_info,
+                )
+            })?;
```
Comment (gelash) [inline api/src/transactions.rs:1204]: I was trying to reason about how the early return changes the semantics of this method (e.g. there is exe_status below but it goes according to output.status, there is the other place you removed where vm_status is used, but also the fact that the rest of the code won't even be performed), can you help me out?

File: api/src/tests/transactions_test.rs [hunk 3]
```diff
@@ -1387,19 +1387,15 @@ async fn simulation_failure_error_message() {
...
             "bytecode": "a11ceb0b030000000105000100000000050601000000000000000600000000000000001a0102",
         },
         "type_arguments": [],
         "arguments": [],
-    }), 200).await;
-
-    assert!(output.as_array().unwrap()[0]["vm_status"]
-        .as_str()
-        .unwrap()
-        .contains("Division by zero"));
```
Comment (gelash) [inline api/src/tests/transactions_test.rs:1402]: is this contains still checked?

File: api/src/tests/multisig_transactions_test.rs [hunk 4]
```diff
@@ -385,11 +385,10 @@ async fn test_multisig_transaction_simulation() {
             "0x1::aptos_account::transfer",
             &[],
             &[&owner_account_1.address().to_hex_literal(), "2000"],
-            200,
+            500,
         )
         .await;
-    let simulation_resp = &simulation_resp.as_array().unwrap()[0];
-    assert!(!simulation_resp["success"].as_bool().unwrap());
```
Comment (gelash) [inline api/src/tests/multisig_transactions_test.rs:392]: is this still checked?

Comment (gelash) [reply api/src/tests/multisig_transactions_test.rs:392]: I see, now it's checked with golden.

File: api/src/tests/transactions_test.rs (hunk 3 above)
Comment (gelash) [reply api/src/tests/transactions_test.rs:1402]: I see golden again

File: api/goldens/aptos_api__tests__simulation_test__test_simulate_transaction_with_valid_signature.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+  "message": "Simulated transactions must have a non-valid signature",
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply api/goldens/aptos_api__tests__simulation_test__test_simulate_transaction_with_valid_signature.json:2]: This is when validating submitted transaction, adjusted!

File: api/src/transactions.rs [hunk 6]
```diff
@@ -1233,22 +1245,7 @@ impl TransactionsApi {
                 let mut user_transactions = Vec::new();
                 for transaction in transactions.into_iter() {
                     match transaction {
-                        Transaction::UserTransaction(user_txn) => {
-                            let mut txn = *user_txn;
-                            match &vm_status {
-                                VMStatus::Error {
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply api/src/transactions.rs:1239]: Yep, and let's see what @davidiw @movekevin and @runtian-zhou think, can change it back so no early returns and instead we record such executions as still sucessful.

File: api/src/tests/transactions_test.rs (hunk 3 above)
Comment (runtian-zhou) [reply api/src/tests/transactions_test.rs:1402]: Hmm where in the golden file? Didn't see this message.

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 1 above)
Comment (runtian-zhou) [reply aptos-move/aptos-vm/src/aptos_vm.rs:1225]: Sry this looks good to me!

File: api/src/transactions.rs (hunk 2 above)
Comment (georgemitenkov) [reply api/src/transactions.rs:1204]: So here we don't return a JSON with transaction information, such as events, gas, etc. I have basically replaced `Result<(VMStatus, VMOutput), VMStatus>` with `Result<VMOutput, VMStatus>` which might not be desired actually... I think I will actually return the old flow, good point about that, thanks!

**Review by runtian-zhou — APPROVED** [review]
Thank you!

File: api/src/tests/transactions_test.rs (hunk 3 above)
Comment (georgemitenkov) [reply api/src/tests/transactions_test.rs:1402]: Copy paste gone wrong, updated, thanks for catching!

File: api/src/transactions.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply api/src/transactions.rs:1239]: Also cc @banool

[the remark this answers is not in the document]
Comment (georgemitenkov) [reply api/src/transactions.rs]: Update: old flow is preserved, simulation can now result in
1. Failure, e.g. when we have a valid signature or fail applying deltas.
2. Success, with failed `VMStatus` and an error message.
3. Success, with success `VMStatus`.

File: api/src/transactions.rs [hunk 7]
```diff
@@ -1183,15 +1183,25 @@ impl TransactionsApi {
...
             ));
         }
 
         // Simulate transaction
         let state_view = self.context.latest_state_view_poem(&ledger_info)?;
-        let (vm_status, output) = AptosVM::simulate_signed_transaction(&txn, &state_view);
+        let vm = AptosVM::new(&state_view.as_move_resolver()).for_simulation();
+
+        let (vm_status, output) =
+            vm.simulate_signed_transaction(&txn, &state_view)
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply api/src/transactions.rs:1197]: I had an assert before but then changed to error after reviews🤣 Switching back to assert inside `simulate_signed_transaction`.

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 8]
```diff
@@ -1929,149 +1990,28 @@ impl VMValidator for AptosVM {
...
-            MemoryTrackedGasMeter::new(StandardGasMeter::new(StandardGasAlgebra::new(
-                self.0.vm_impl.get_gas_feature_version(),
-                gas_params.vm.clone(),
-                storage_gas_params.clone(),
-                txn_data.max_gas_amount(),
-            )));
+        transaction: &SignedTransaction,
+        state_view: &impl StateView,
+    ) -> Result<(VMStatus, TransactionOutput), VMStatus> {
+        assert!(self.is_simulation, "VM has to be created for simulation");
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/aptos_vm.rs:1999]: Isn't this case similar to signature verification? The caller checks signature is not valid and creates a VM for simulation, this is just a callee, so I used asserts. Feels like a developer responsibility to create the VM and set the right flag?

Let me know if you prefer the error, I ll change this (though I have just changed signature check to assertion back 😂)

File: aptos-move/aptos-vm/src/lib.rs [hunk 9]
```diff
@@ -169,11 +169,12 @@ pub trait VMExecutor: Send + Sync {
...
 
-/*
-/// Get the AccessPath to a resource stored under `address` with type name `tag`
-/// DNS
-fn create_access_path(address: AccountAddress, tag: StructTag) -> anyhow::Result<AccessPath> {
-    let resource_tag = ResourceKey::new(address, tag);
-    AccessPath::resource_access_path(resource_tag)
+/// This trait describes the VM's simulation interfaces.
+pub trait VMSimulator {
+    /// Executes a SignedTransaction without performing signature verification.
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/lib.rs:174]: Result is due to V1 aggregator. I guess it should never happen, adjusted.

File: api/src/transactions.rs (hunk 2 above)
Comment (georgemitenkov) [reply api/src/transactions.rs]: Actually now it is `(VMStatus, VMOutput)`, because made the verification of signature an assert and aggregator V1 deltas to never fail.

File: api/src/tests/simulation_test.rs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    let resp = simulate_aptos_transfer(&mut context, true, 400).await;
+    context.check_golden_output(resp);
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+async fn test_simulate_transaction_with_invalid_signature() {
+    let mut context = new_test_context(current_function_name!());
+    let resp = simulate_aptos_transfer(&mut context, false, 200).await;
+    assert!(resp[0]["success"].as_bool().is_some_and(|v| v));
+}
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply api/src/tests/simulation_test.rs:77]: added a test

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm/src/aptos_vm.rs:2000]: @zekun000 @igor-aptos @vgao1996 
A concrete question here regarding asserts vs errors:
call-site makes sure signatures are valid & vm is_simulation, but the caller is in a different crate. Should we have an assert here, or should we return an error?

@georgemitenkov: as you mentioned, for non-simulation flow, let's also add a check that the vm isn't in simulation mode. Let's do the same convention (assert or result) as above.

**Review by gelash — APPROVED** [review]

**Comment by georgemitenkov** [comment]
Replay passes: https://github.com/aptos-labs/aptos-core/actions/runs/6962079009, merging then
